### PR TITLE
3-jax-mlp: MLP Model in Jax with Metrics 

### LINKED ISSUE #4: Jax - MLP

### PR DESCRIPTION
**Pull request by UrbanPistek**
Under `multi-layer-perceptron/jax` you will find the following: 

- `jax_notebook.ipynb` A notebook with a jax overview 
- `jax_test.py` with which trains the model and computes the metrics, with a module to define the mlp model and provide utility functions to train a model in jax
- `env.yml` which defines the environment to use for this project 
- `README.md` contains some instructions on how to run things in this project 

**To run the main script to train and evaluate the MLP model**
```
conda env create --file env.yml
conda activate jax-cpu
python -u jax_test.py
``` 

**Note:** I am gonna make some changes to better cleanup the code or refactor things to be a bit cleaner

### CODE REVIEW

File: multi-layer-perceptron/jax/jax_overview.ipynb [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+    "    return scale * random.normal(w_key, (n, m)), scale * random.normal(b_key, (n,))\n",
+    "\n",
+    "# Initialize all layers for a fully-connected neural network with sizes \"sizes\"\n",
+    "def init_network_params(sizes, key):\n",
+    "    keys = random.split(key, len(sizes))\n",
+    "    return [random_layer_params(m, n, k) for m, n, k in zip(sizes[:-1], sizes[1:], keys)]\n",
+    "\n",
+    "# Hyperparameters\n",
+    "# layer_sizes = [784, 512, 512, 10]\n",
+    "layer_sizes = [784, 512, 512, 10]\n",
```
Comment (Anu2711) [inline multi-layer-perceptron/jax/jax_overview.ipynb]: The hidden layer size should be 100 not 512

File: multi-layer-perceptron/jax/main.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    print(f"\nEnvironment Config: ")
+    print(f"TF_CPP_MIN_LOG_LEVEL = {os.environ['TF_CPP_MIN_LOG_LEVEL']}")
+
+    # Initialize model
+    layers = [784, 100, 10]
+    batch_size = 128
+    epochs = 10
+    n_targets = 10
+    learning_rate = 0.01
```
Comment (trevor-yu-087) [inline multi-layer-perceptron/jax/main.py]: Spec says 1e-4 but probably gonna change it to 1e-3 during the meeting.

File: multi-layer-perceptron/jax/mlp.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    activations = input
+    for w, b in params[:-1]:
+        outputs = jnp.dot(w, activations) + b
+        activations = relu(outputs)
+    
+    final_w, final_b = params[-1]
+    logits = jnp.dot(final_w, activations) + final_b
+    
+    return logits - logsumexp(logits)
```
Comment (trevor-yu-087) [inline multi-layer-perceptron/jax/mlp.py]: Maybe leave a comment saying this is log softmax loss

File: multi-layer-perceptron/jax/mlp.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+# Temp, move this
+batched_predict = vmap(predict, in_axes=(None, 0))
+
+def loss(params, inputs, targets):
+    """
+    Compute loss
```
Comment (trevor-yu-087) [inline multi-layer-perceptron/jax/mlp.py]: This loss would be negative log liklihood loss, add a comment saying this is what it's computing.

**Comment by trevor-yu-087** [comment]
As a suggestion, there's also the Haiku and Flax API built on top of Jax for implementing common neural network primitives like dense layers, optimizers, etc. You might want to consider using those in the future so you don't need to create all the primitives from scratch.

File: multi-layer-perceptron/jax/mlp.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    """
+    preds = batched_predict(params, inputs)
+    loss = -jnp.mean(preds * targets)
+
+    return loss
+
+@jit
+def update(params, x, y, learning_rate):
```
Comment (trevor-yu-087) [inline multi-layer-perceptron/jax/mlp.py]: This is simple stochastic gradient descent, not Adam. It's kind of a pain to implement Adam from scratch (although doable). Might want to use an existing optimizer in Flax or Haiku.

Comment (functionstackx) [reply multi-layer-perceptron/jax/mlp.py]: have yall looked into optax?

Comment (UrbanPistek) [reply multi-layer-perceptron/jax/mlp.py]: Yeah, looking into it I think I have seomthing I can use to implement adam relitively quickly. I think I'll do that first, then I might look into what it would take to implement adam from scratch as it might be a good exercise I think.

File: multi-layer-perceptron/jax_test.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        # Main training loop
+        for x, y in ds:
+            x = jnp.reshape(x, (len(x), num_pixels))
+            y = one_hot_encoder(y, num_labels)
+            params = update(params, x, y, learning_rate)
+        
+        epoch_time = time.time() - start_time
+        epoch_times.append(epoch_time)
+
+        train_acc = accuracy(params, train_images, train_labels)
```
Comment (trevor-yu-087) [inline multi-layer-perceptron/jax_test.py]: For consistency with other implementation, exclude this call to accuracy here since it'll do another whole loop through the training data to do inference. Most of the others do the accuracy prediction at the same time as processing the batch during the update step and having another run through the training data will add time to each epoch.

File: multi-layer-perceptron/jax_test.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    return loss
+
+@jit
+def update(params, x, y, learning_rate):
+    """
+    Perform backpropagation using stochastic gradient descent.
+    """
+    grads = grad(loss)(params, x, y)
+
+    return [(w - learning_rate * dw, b - learning_rate * db) for (w, b), (dw, db) in zip(params, grads)]
```
Comment (trevor-yu-087) [inline multi-layer-perceptron/jax_test.py:144]: SGD still, what did you decide to do about Adam? Write your own or use the implementation from Flax or Haiku?

File: multi-layer-perceptron/jax_test.py (hunk 6 above)
Comment (UrbanPistek) [reply multi-layer-perceptron/jax_test.py]: Moved so that accuracy is only called is debug is enabled, so no accuarcy by default.

File: multi-layer-perceptron/jax_test.py (hunk 7 above)
Comment (UrbanPistek) [reply multi-layer-perceptron/jax_test.py:144]: Implementat adam using optax - essential another packages built of off jax for optimizers. 

RIght now the script is configured so that it can easily run either gradient descent or adam, with adam being the default its setup for. 

I think going forward I will look into implementing adam in jax from scratch to be more inline with the ground up implementation and style of Jax, but for now I am gonna have it like this so its wrapped up.

**Review by trevor-yu-087 — APPROVED** [review]
